pkg/controllers/image/image.go

```diff
@@ -0,0 +1,338 @@
+	"github.com/rancher/fleet/pkg/apis/fleet.cattle.io/v1alpha1"
+	fleetcontrollers "github.com/rancher/fleet/pkg/generated/controllers/fleet.cattle.io/v1alpha1"
+	corev1controler "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/rancher/wrangler/pkg/kstatus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	lock sync.Mutex
+)
+
+const (
+	// AlphabeticalOrderAsc ascending order
+	AlphabeticalOrderAsc = "ASC"
+	// AlphabeticalOrderDesc descending order
+	AlphabeticalOrderDesc = "DESC"
+
+	defaultMessageTemplate = `Update from image update automation`
+)
+
+func Register(ctx context.Context, core corev1controler.Interface, gitRepos fleetcontrollers.GitRepoController, images fleetcontrollers.ImageScanController) error {
+	h := handler{
+		secretCache:  core.Secret().Cache(),
+		gitrepoCache: gitRepos.Cache(),
+	}
+
+	fleetcontrollers.RegisterImageScanStatusHandler(ctx, images, "ImageScanned", "image-scan", h.onChange)
+	return nil
+}
+
+type handler struct {
+	secretCache  corev1controler.SecretCache
+	gitrepoCache fleetcontrollers.GitRepoCache
+}
+
+func (h handler) onChange(image *v1alpha1.ImageScan, status v1alpha1.ImageScanStatus) (v1alpha1.ImageScanStatus, error) {
+	if image == nil || image.DeletionTimestamp != nil {
+		return status, nil
+	}
+
+	if image.Spec.Suspend {
+		return status, nil
+	}
+
+	ref, err := name.ParseReference(image.Spec.Image)
+	if err != nil {
+		kstatus.SetError(image, err.Error())
+		return status, err
+	}
+
+	canonical := ref.Context().String()
+	if canonical != status.CanonicalImageName {
+		status.CanonicalImageName = canonical
+	}
+
+	if !shouldScan(image) {
+		return status, nil
+	}
+
+	var options []remote.Option
+	if image.Spec.SecretRef != nil {
+		secret, err := h.secretCache.Get(image.Namespace, image.Spec.SecretRef.Name)
+		if err != nil {
+			kstatus.SetError(image, err.Error())
+			return status, err
+		}
+		auth, err := authFromSecret(secret, ref.Context().RegistryStr())
+		if err != nil {
+			kstatus.SetError(image, err.Error())
+			return status, err
+		}
+		options = append(options, remote.WithAuth(auth))
+	}
+
+	tags, err := remote.ListWithContext(context.Background(), ref.Context(), options...)
+	if err != nil {
+		kstatus.SetError(image, err.Error())
+		return status, err
+	}
+
+	status.LastScanTags = tags
+	status.LastScanTime = metav1.NewTime(time.Now())
+
+	latestTag, err := latestTag(image.Spec.Policy, tags)
+	if err != nil {
+		kstatus.SetError(image, err.Error())
+		return status, err
+	}
+
+	status.LatestTag = latestTag
+	status.LatestImage = status.CanonicalImageName + ":" + latestTag
+
+	if image.Spec.GitRepoName == "" {
+		return status, nil
+	}
+
+	gitrepo, err := h.gitrepoCache.Get(image.Namespace, image.Spec.GitRepoName)
+	if err != nil {
+		kstatus.SetError(image, err.Error())
+		return status, err
+	}
+
+	lock.Lock()
+	defer lock.Unlock()
+	// todo: maybe we should preserve the dir
```

Comment:
What would be the advantage of preserving the temporary directory? In case of error?

Reply:
So it doesn't have to clone from scatch.